### PR TITLE
Ch. 1: Issue 175:  Separate out the PMIx client interface from the CL/host interactions 

### PR DESCRIPTION
**Pull request by dsolt**
…y searching

### CODE REVIEW

**Comment by jjhursey** [comment]
Ref Issue #175

File: Chap_Introduction.tex [hunk 1]
```diff
@@ -2,135 +2,67 @@
...
+system software components providing them, the 
+community represented by the Administrative Steering Committee (\ac{ASC})
+feels that the development of a shared specification better serves the
+community.  Such implementations can interface
+with existing system software that are already compliant with \ac{PMIx} or, 
+when not already available, directly implement PMIx APIs.
+As a result, \ac{PMIx} enables programming languages and libraries to focus on their core
+competencies without having to provide their own system-level services.
+
+\subsection{Background}
```
Comment (jjhursey) [inline Chap_Introduction.tex]:
```suggestion
\section{Background}
```

File: Chap_Introduction.tex [hunk 2]
```diff
@@ -162,52 +91,70 @@ \section{PMIx Architecture Overview}
...
+authenticate with the \ac{PMIx} server.  
+The \ac{PMIx} server will have sufficient knowledge about the  
+process which it created, either directly or through other \ac{SMS}, to authenticate the 
+process and provide information the process requests such as its identity and the 
+identity of its peers.  
+
+Tools, whether standalone or embedded in job scripts, are an exception to 
+the normal client registration process.   A process can register as a tool, provided 
+the \ac{PMIx} client library has adequate rendezvous information to connect to the appropriate
+\ac{PMIc} server.  This allows processes
```
Comment (jjhursey) [inline Chap_Introduction.tex]:
```suggestion
\ac{PMIx} server.  This allows processes
```

**Comment by jjhursey** [comment]
@dsolt I have a couple of inline fixes for a section header and `ac` expansion above that I didn't see reflected in the current version of the document.

File: Chap_Introduction.tex [hunk 3]
```diff
@@ -143,71 +74,112 @@ \section{PMIx Architecture Overview}
...
+the normal client registration process.   A process can register as a tool, provided 
+the \ac{PMIx} client library has adequate rendezvous information to connect to the appropriate
+\ac{PMIx} server.  This allows processes
+which were not created by the PMIx infrastructure to request access to PMIx functionality. 
+Processes registered as tools do not have peers.  
+
+\section{Portability of Functionality}
+\label{chap:intro:not_supported}
+
+It is difficult to define a portable \ac{API} that will provice access to the many
```
Comment (jjhursey) [inline Chap_Introduction.tex]:
```suggestion
It is difficult to define a portable \ac{API} that will provide access to the many             
```

**Comment by jjhursey** [comment]
We had a few notes from the Q4 2019 meeting for this document that we should make sure are reviewed:
 * https://github.com/pmix/pmix-standard/wiki/ASC-Q4-2019-Meeting

File: Chap_Introduction.tex (hunk 3 above)
Comment (dsolt) [reply Chap_Introduction.tex]: done

File: Chap_Introduction.tex (hunk 2 above)
Comment (dsolt) [reply Chap_Introduction.tex]: done

File: Chap_Introduction.tex [hunk 4]
```diff
@@ -143,71 +74,112 @@ \section{PMIx Architecture Overview}
...
+
+As clients invoke \ac{PMIx} \acp{API}, is possible that some client requests can 
+be handled at the client level.  Other requests might require communication with the 
+local \ac{PMIx} server, which subsequently might request services from the host \ac{SMS} 
+(represented here by an \ac{RM} daemon).  
+
+The interaction between the \ac{PMIx} server and \ac{SMS} might be 
+achieved using callback functions registered during server initialization.  
+The host \ac{SMS} could indicate its lack of support for any operation by simply providing a \textit{NULL} for the associated callback function, or could create a function entry that returns \textit{not supported} when called.  These interactions between a \ac{PMIx} implementation 
+and the \ac {SMS} are not defined by the \ac {PMIx} standard.
```
Comment (SteVwonder) [inline Chap_Introduction.tex]: I thought the `not supported` case is defined by the PMIx standard.

File: Chap_Introduction.tex [hunk 5]
```diff
@@ -143,71 +74,112 @@ \section{PMIx Architecture Overview}
...
+identity of its peers.  
+
+As clients invoke \ac{PMIx} \acp{API}, is possible that some client requests can 
+be handled at the client level.  Other requests might require communication with the 
+local \ac{PMIx} server, which subsequently might request services from the host \ac{SMS} 
+(represented here by an \ac{RM} daemon).  
+
+The interaction between the \ac{PMIx} server and \ac{SMS} might be 
+achieved using callback functions registered during server initialization.  
+The host \ac{SMS} could indicate its lack of support for any operation by simply providing a \textit{NULL} for the associated callback function, or could create a function entry that returns \textit{not supported} when called.  These interactions between a \ac{PMIx} implementation
```
Comment (SteVwonder) [inline Chap_Introduction.tex]: Are these callback functions for the server-side interface?

File: Chap_Introduction.tex (hunk 4 above)
Comment (dsolt) [reply Chap_Introduction.tex]: yes, i think that is outdated.  the most recent commits have put this back to show that it is defined

File: Chap_Introduction.tex (hunk 5 above)
Comment (dsolt) [reply Chap_Introduction.tex]: see above

**Comment by dsolt** [comment]
[Text_changes_NEW-to-ORIG.docx](https://github.com/pmix/pmix-standard/files/4091772/Text_changes_NEW-to-ORIG.docx)

**Comment by dsolt** [comment]
[Text_changes_NEW-to-ORIG.docx](https://github.com/pmix/pmix-standard/files/4104993/Text_changes_NEW-to-ORIG.docx)
[Text_changes_ORIG-to-NEW.docx](https://github.com/pmix/pmix-standard/files/4104995/Text_changes_ORIG-to-NEW.docx)

**Comment by jjhursey** [comment]
PMIx ASC Q1'2020 Meeting:
 * The reading generally went well. A number of changes which were mostly minor in nature were recommended. A couple of recommendations could become significant in size/implications depending on how the working group handles them. As such, the group decided to neither officially push back nor vote on this proposal in this meeting.
 * The group suggested the following:
   * The working group should apply the recommended changes to Chapter 1
   * Post a new version of the document to this PR with a reference to the commit(s) in which they were implemented.
   * Post a new straw poll on this PR with a note to the mailing list.
   * ASC members will vote in the straw poll and note in the comments if they wish to have a full formal reading in the next meeting. Lack of such a comment will indicate that the changes were minor enough to not warrant a full reading.
   * If a full reading is requested then it will be scheduled in the next meeting with a possible first formal vote immediately afterward.
   * If a full reading is not requested then a summary reading of the changes will be presented in the next meeting followed by the first vote.


A summary of the recommended changes are indicated in the [notes](https://github.com/pmix/pmix-standard/wiki/ASC-Q1-2020-Meeting) from the ASC meeting.

File: Chap_Introduction.tex [hunk 6]
```diff
@@ -143,71 +74,121 @@ \section{PMIx Architecture Overview}
...
-any PMIx server providing they are given adequate rendezvous information. The PMIx conceptual model views the
-collection of PMIx servers as a cloud-like conglomerate --- i.e., orchestration and information requests can be
-given to any server regardless of location. However, tools frequently execute on locations that may not house an
-operating PMIx server --- e.g., a users notebook computer. Thus, tools need the ability to remotely connect to
-the PMIx server ``cloud''.
+Fig.~\ref{fig:roles} shows how tools can interact with the \ac{PMIx} architecture.
+Tools, whether standalone or embedded in job scripts, are an exception to 
+the normal client registration process.   A process can register as a tool, provided 
+the \ac{PMIx} client library has adequate rendezvous information to connect to the appropriate
+\ac{PMIx} server.  This allows processes
```
Comment (jjhursey) [inline Chap_Introduction.tex]:
```suggestion
\ac{PMIx} server (either hosted on the local machine or on a remote machine).  This allows processes
```

File: Chap_Introduction.tex [hunk 7]
```diff
@@ -143,71 +74,121 @@ \section{PMIx Architecture Overview}
...
+local \ac{PMIx} server, subsequently might request services from the host \ac{SMS} 
+(represented here by a \ac{RM} daemon).  
+The interaction between the \ac{PMIx} server and \ac{SMS} are 
+achieved using callback functions registered during server initialization.  
+The host \ac{SMS} can indicate its lack of support for any operation by simply providing a \textit{NULL} for the associated callback function, or can create a function entry that returns \textit{not supported} when called.
 
 Recognizing the burden this places on SMS vendors, the PMIx community has included interfaces by
-which the host can request support from local SMS elements. Once the SMS has transferred the request to
-an appropriate location, a PMIx server interface can be used to pass the request between SMS subsystems.
+which the host can request support from local SMS elements. Once the \ac{SMS} has transferred the request to
```
Comment (jjhursey) [inline Chap_Introduction.tex]:
```suggestion
which the host \ac{SMS} (containing the local PMIx service instance) can request support from local \ac{SMS} elements via the PMIx API. Once the \ac{SMS} has transferred the request to
```

File: Chap_Introduction.tex [hunk 8]
```diff
@@ -143,71 +74,121 @@ \section{PMIx Architecture Overview}
...
-given to any server regardless of location. However, tools frequently execute on locations that may not house an
-operating PMIx server --- e.g., a users notebook computer. Thus, tools need the ability to remotely connect to
-the PMIx server ``cloud''.
+Fig.~\ref{fig:roles} shows how tools can interact with the \ac{PMIx} architecture.
+Tools, whether standalone or embedded in job scripts, are an exception to 
+the normal client registration process.   A process can register as a tool, provided 
+the \ac{PMIx} client library has adequate rendezvous information to connect to the appropriate
+\ac{PMIx} server.  This allows processes
+which were not created by the PMIx infrastructure to request access to PMIx functionality. 
+Processes registered as tools do not have peers.
```
Comment (jjhursey) [inline Chap_Introduction.tex]: Per discussion on teleconf - strike this last sentence. It _may_ be possible that tools have peers (see `PMIX_TOOL_NSPACE` / `PMIX_TOOL_RANK` in the `PMIx_tool_init` call)

File: Chap_Introduction.tex [hunk 9]
```diff
@@ -88,10 +88,10 @@ \section{PMIx Architecture Overview}
 process and provide information the process requests such as its identity and the 
 identity of its peers.  
 
-As clients invoke \ac{PMIx} \acp{API}, is possible that some client requests can 
+As clients invoke \ac{PMIx} \acp{API}, it is possible that some client requests can 
 be handled at the client level.  Other requests might require communication with the 
-local \ac{PMIx} server, which subsequently might request services from the host \ac{SMS} 
-(represented here by an \ac{RM} daemon).  
+local \ac{PMIx} server, subsequently might request services from the host \ac{SMS}
```
Comment (jjhursey) [inline Chap_Introduction.tex]:
```suggestion
local \ac{PMIx} server, which subsequently might request services from the host \ac{SMS} 
```

File: Chap_Introduction.tex [hunk 10]
```diff
@@ -143,71 +74,121 @@ \section{PMIx Architecture Overview}
...
+particular function and allow client applications to query and adjust to the level of support available.  \ac{PMIx} clients should be written to account for the possibility that a \ac{PMIx} \ac{API} might return an error code indicating that the call is not supported.
+The \ac{PMIx} community continues to look at ways to assist \ac{SMS} implementers in their decisions 
+on what functionality to support by highlighting functions and attributes that are 
+critical to basic application execution (e.g., \refapi{PMIx_Get}) for certain classes of applications.
+
+\subsection{Optional Nature of Attributes}
+
+An area where differences between support on different systems can be challenging is regarding the attributes that provide information to the client process and/or control the behavior of a \ac{PMIx} \ac{API}.  Most
+\ac{PMIx} \ac{API} calls can accept additional information or attributes specified in the form of 
+key/value pairs. These attributes provide information to the \ac{PMIx} implementation which influence the
```
Comment (jjhursey) [inline Chap_Introduction.tex]:
```suggestion
key/value pairs. These attributes provide information to the \ac{PMIx} implementation that influence the 
```

**Comment by dsolt** [comment]
We would like feedback on whether the changes in the last commit are sufficiently trivial to avoid a moving this ticket back to a first reading.  If you are ok with us moving this forward to a 2nd reading  please emoji a thumbs up.  If you prefer that the working group represent this as a first reading because you feel the changes are significant, please emoji a thumbs down.

Please use emoji reactions ON THIS COMMENT to indicate your position on this proposal.

* You do not need to vote on every proposal
* If you have no opinion, don't vote - that is also useful data
* If you've already commented on this issue, please still vote so

File: Chap_Introduction.tex [hunk 11]
```diff
@@ -143,71 +74,120 @@ \section{PMIx Architecture Overview}
...
+process which it created, either directly or through other \ac{SMS}, to authenticate the 
+process and provide information the process requests such as its identity and the 
+identity of its peers.  
+
+As clients invoke \ac{PMIx} \acp{API}, it is possible that some client requests can 
+be handled at the client level.  Other requests might require communication with the 
+local \ac{PMIx} server, which subsequently might request services from the host \ac{SMS} 
+(represented here by a \ac{RM} daemon).  
+The interaction between the \ac{PMIx} server and \ac{SMS} are 
+achieved using callback functions registered during server initialization.
```
Comment (abouteiller) [inline Chap_Introduction.tex:96]: An SMS may choose to completely embed the PMIx server; thus not needing the PMIx server API. 

Maybe different wording would work better;
```suggestion
To facilitate the interaction between the \ac{PMIx} server and the \ac{SMS}, the \ac{PMIx} server API provides callback functions to be registered during server initialization.
```

File: Chap_Introduction.tex [hunk 12]
```diff
@@ -143,71 +74,120 @@ \section{PMIx Architecture Overview}
...
+is not available.
 
-The scope of the PMIx standard therefore spans the range of these interactions, between client-and-SMS and between SMS
-subsystems. Note again that this does not impose a requirement on any given PMIx implementation to cover the entire
-range --- implementers are free to return \textit{not supported} from any PMIx function.
-
-
-%%%%%%%%%%%
-\subsection{The \acf{PRI}}
+It is therefore critical that users and application implementers:
```
Comment (abouteiller) [inline Chap_Introduction.tex:169]: Should start with a definition of what is the PRI

**Comment by jjhursey** [comment]
PMIx ASC 2Q 2020 Meeting:
 * **Passed first vote**: 10 Yes / 0 No / 0 Abstain / 4 not present for vote
 * Discussion point [here](https://github.com/pmix/pmix-standard/pull/192#discussion_r408989451) will be resolved in a separate, following PR.

**Comment by rhc54** [comment]
Given the status of this PR and that I will need to modify Chapter 1 to update the history for release of v4, I would like to commit this into the master branch. Can we discuss and perhaps get agreement to do so at the July meeting given that there doesn't appear to be any opposition?

**Comment by jjhursey** [comment]
PMIx ASC 3Q 2020 Meeting:
 * **Passed Second Vote** 8 Yes / 0 No / 2 Abstain / 4 not present for vote
 * The ASC also discussed including this in the v4 release and had no objection in doing so. So we will merge this into `master` and it will be included in the v4 and future releases.

**Pull request merged**